app/models/user.rb

```diff
@@ -395,6 +395,19 @@ def max_friends_per_hour
     max_friends.clamp(0, Settings.max_friends_per_hour)
   end
 
+  def max_changeset_comments_per_hour
+    if moderator?
+      36000
+    else
+      previous_comments = changeset_comments.limit(200).count
+      active_reports = issues.with_status(:open).sum(:reports_count)
+      max_comments = previous_comments / 200.0 * Settings.max_changeset_comments_per_hour
+      max_comments = max_comments.floor.clamp(Settings.min_changeset_comments_per_hour, Settings.max_changeset_comments_per_hour)
+      max_comments /= 2**active_reports
+      max_comments.floor.clamp(1, Settings.max_changeset_comments_per_hour)
+    end
```

Comment:
That's an interesting algorithm! I do like relying on previous activity, rather than account age, so I think we could rework the existing limits along these lines too. And using reports as the main downward modifier is smart.

I suspect there might be non-moderator power users caught out by this, since a disagreement about reverts could lead to one person making ~10 reports and therefore severely rate-limiting the power user. Perhaps defending against that would be too much complexity for now.

Is there a naming issue here? The minimum number of comments per hour is actually 1, not `min_changeset_comments_per_hour`. I think an admin seeing the min and max settings might not realise that the min is more of a "default" or starter value, that can go both up and also go down lower.

I also think we need to move these rate limit calculations out of the user model, but I'm not yet sure where they should go. The model is already huge, and it almost feels like these are controller-level decisions rather than some property of the model in isolation.

Reply:
The minimum value is the minimum that will happen in normal operation - it will only go below that when the user is reported. Open to suggestions for alternate names though - the basic idea was that the limit ramps up from min to max but can be restricted by reports.

Obviously there is a potential issue with non-moderator power users as you say, though to paraphrase @woodpeck do we really want those people commenting on thousands of changesets and generating email storms?

Reply:
> Obviously there is a potential issue with non-moderator power users as you say, though to paraphrase @woodpeck do we really want those people commenting on thousands of changesets and generating email storms?

:smile: Although if good stuff can be done outside of DWG, more power to them! But in this case, a power user can be prevented from making more than `1` changeset comment per hour, by only a few (bogus) reports.

For naming, my suggestion would be:
* `min_changeset_comments_per_hour` = 1
* `default_changeset_comments_per_hour` = 6
* `max_changeset_comments_per_hour` = 200
* `moderator_changeset_comments_per_hour` = 36_000

Reply:
> But in this case, a power user can be prevented from making more than 1 changeset comment per hour, by only a few (bogus) reports.

Maybe like proposed in https://github.com/openstreetmap/openstreetmap-website/issues/2342#issuecomment-1694650004 it would be possible to count how many users made reports against account, not simply how many reports were made? So simply making 100 bogus reports will not have so serious effects.